sklearn/tree/_criterion.pyx

```diff
@@ -268,6 +312,12 @@ cdef class ClassificationCriterion(Criterion):
                 self.sum_right == NULL):
             raise MemoryError()
 
+        if allow_missing:
+            self.sum_missing = <double*> calloc(n_elements, sizeof(double))
+            self.sum_available = <double*> calloc(n_elements, sizeof(double))
+            if self.sum_missing == NULL or self.sum_available == NULL:
+                raise MemoryError()
```

Comment:
In my opinion, we should allocate arrays in a single way only, to avoid easily overlooked bugs when dealing with such things. This is why at some point we wrote `safe_realloc` and replaced all these calls. 

That is:

```
safe_realloc(&self.sum_missing, n_elements)
safe_realloc(&self.sum_available, n_elements)
```